File: src/main/java/org/cubexmc/metro/gui/GuiHolder.java
package org.cubexmc.metro.gui;

import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.InventoryHolder;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * GUI 持有者，用于标识和存储 GUI 数据
 */
public class GuiHolder implements InventoryHolder {
    
    public enum GuiType {
        MAIN_MENU,      // 主菜单
        LINE_LIST,      // 线路列表
        STOP_LIST,      // 站点列表
        LINE_VARIANTS,  // 线路变体列表
        STOP_VARIANTS,  // 站点变体列表
        LINE_DETAIL,    // 线路详情
        STOP_DETAIL,    // 站点详情
        ADD_STOP_LIST,  // 添加站点列表
        ADD_STOP_VARIANTS, // 添加站点变体列表
        LINE_BOARDING_CHOICE, // 乘车线路选择
        LINE_SETTINGS,  // 线路设置
        STOP_SETTINGS,  // 站点设置
        CONFIRM_ACTION  // 操作确认
    }
    
    private final GuiType type;
    private final Map<String, Object> data;
    private GuiView previousView;
    private Inventory inventory;
    
    public GuiHolder(GuiType type) {
        this.type = type;
        this.data = new HashMap<>();
    }
    
    public GuiType getType() {
        return type;
    }
    
    public void setData(String key, Object value) {
        data.put(key, value);
    }
    
    @SuppressWarnings("unchecked")
    public <T> T getData(String key) {
        return (T) data.get(key);
    }
    
    public <T> T getData(String key, T defaultValue) {
        Object value = data.get(key);
        if (value == null) {
            return defaultValue;
        }
        return (T) value;
    }

    public GuiView getPreviousView() {
        return previousView;
    }

    public void setPreviousView(GuiView previousView) {
        this.previousView = previousView;
    }

    public GuiView snapshot() {
        return new GuiView(type, data, previousView);
    }
    
    public void setInventory(Inventory inventory) {
        this.inventory = inventory;
    }
    
    @Override
    public Inventory getInventory() {
        return inventory;
    }

    public static final class GuiView {
        private final GuiType type;
        private final Map<String, Object> data;
        private final GuiView previousView;

        private GuiView(GuiType type, Map<String, Object> data, GuiView previousView) {
            this.type = type;
            this.data = Collections.unmodifiableMap(new HashMap<>(data));
            this.previousView = previousView;
        }

        public GuiType getType() {
            return type;
        }

        @SuppressWarnings("unchecked")
        public <T> T getData(String key) {
            return (T) data.get(key);
        }

        @SuppressWarnings("unchecked")
        public <T> T getData(String key, T defaultValue) {
            Object value = data.get(key);
            if (value == null) {
                return defaultValue;
            }
            return (T) value;
        }

        public GuiView getPreviousView() {
            return previousView;
        }
    }
}

File: src/main/java/org/cubexmc/metro/gui/GuiHolder.kt
package org.cubexmc.metro.gui

import java.util.Collections
import org.bukkit.inventory.Inventory

/** GUI 持有者，用于标识和存储 GUI 数据。 */
class GuiHolder(private val type: GuiType) : NullableInventoryHolder() {
    enum class GuiType {
        MAIN_MENU,
        LINE_LIST,
        STOP_LIST,
        LINE_VARIANTS,
        STOP_VARIANTS,
        LINE_DETAIL,
        STOP_DETAIL,
        ADD_STOP_LIST,
        ADD_STOP_VARIANTS,
        LINE_BOARDING_CHOICE,
        LINE_SETTINGS,
        STOP_SETTINGS,
        CONFIRM_ACTION,
    }

    private val data: MutableMap<String, Any?> = HashMap()
    private var previousView: GuiView? = null
    private var inventory: Inventory? = null

    fun getType(): GuiType = type

    fun setData(key: String, value: Any?) {
        data[key] = value
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getData(key: String): T? = data[key] as T?

    @Suppress("UNCHECKED_CAST")
    fun <T> getData(key: String, defaultValue: T): T {
        val value = data[key] ?: return defaultValue
        return value as T
    }

    fun getPreviousView(): GuiView? = previousView

    fun setPreviousView(previousView: GuiView?) {
        this.previousView = previousView
    }

    fun snapshot(): GuiView = GuiView(type, data, previousView)

    fun setInventory(inventory: Inventory?) {
        this.inventory = inventory
    }

    override fun currentInventory(): Inventory? = inventory

    class GuiView internal constructor(
        private val type: GuiType,
        data: Map<String, Any?>,
        private val previousView: GuiView?,
    ) {
        private val data: Map<String, Any?> = Collections.unmodifiableMap(HashMap(data))

        fun getType(): GuiType = type

        @Suppress("UNCHECKED_CAST")
        fun <T> getData(key: String): T? = data[key] as T?

        @Suppress("UNCHECKED_CAST")
        fun <T> getData(key: String, defaultValue: T): T {
            val value = data[key] ?: return defaultValue
            return value as T
        }

        fun getPreviousView(): GuiView? = previousView
    }
}
